ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/backwardsync/ForwardSyncPhase.java

```diff
@@ -0,0 +1,266 @@
+          LOG,
+          "Requesting {} blocks {}->{} ({})",
+          blockHeaders::size,
+          () -> blockHeaders.get(0).getNumber(),
+          () -> blockHeaders.get(blockHeaders.size() - 1).getNumber(),
+          () -> blockHeaders.get(0).getHash().toHexString());
+      return requestBodies(blockHeaders).thenApply(this::saveBlocks);
+    }
+  }
+
+  @VisibleForTesting
+  protected CompletableFuture<Block> requestBlock(final BlockHeader blockHeader) {
+    final GetBlockFromPeerTask getBlockFromPeerTask =
+        GetBlockFromPeerTask.create(
+            context.getProtocolSchedule(),
+            context.getEthContext(),
+            Optional.of(blockHeader.getHash()),
+            blockHeader.getNumber(),
+            context.getMetricsSystem());
+    final CompletableFuture<AbstractPeerTask.PeerTaskResult<Block>> run =
+        getBlockFromPeerTask.run();
+    return run.thenApply(AbstractPeerTask.PeerTaskResult::getResult);
+  }
+
+  @VisibleForTesting
+  protected CompletableFuture<List<Block>> requestBodies(final List<BlockHeader> blockHeaders) {
+    final GetBodiesFromPeerTask getBodiesFromPeerTask =
+        GetBodiesFromPeerTask.forHeaders(
+            context.getProtocolSchedule(),
+            context.getEthContext(),
+            blockHeaders,
+            context.getMetricsSystem());
+
+    final CompletableFuture<AbstractPeerTask.PeerTaskResult<List<Block>>> run =
+        getBodiesFromPeerTask.run();
+    return run.thenApply(AbstractPeerTask.PeerTaskResult::getResult)
+        .thenApply(
+            blocks -> {
+              blocks.sort(Comparator.comparing(block -> block.getHeader().getNumber()));
+              return blocks;
+            });
+  }
+
+  @VisibleForTesting
+  protected Void saveBlock(final Block block) {
+    debugLambda(LOG, "Going to validate block {}", () -> block.getHeader().getHash().toHexString());
+    var optResult =
+        context
+            .getBlockValidator(block.getHeader().getNumber())
+            .validateAndProcessBlock(
+                context.getProtocolContext(),
+                block,
+                HeaderValidationMode.FULL,
+                HeaderValidationMode.NONE);
+
+    optResult.blockProcessingOutputs.ifPresent(
+        result -> {
+          debugLambda(
+              LOG,
+              "Block {} was validated, going to import it",
+              () -> block.getHeader().getHash().toHexString());
+          result.worldState.persist(block.getHeader());
+          context.getProtocolContext().getBlockchain().appendBlock(block, result.receipts);
+        });
+    return null;
+  }
+
+  @VisibleForTesting
+  protected Void saveBlocks(final List<Block> blocks) {
+
+    for (Block block : blocks) {
+      final Optional<Block> parent =
+          context
+              .getProtocolContext()
+              .getBlockchain()
+              .getBlockByHash(block.getHeader().getParentHash());
+      if (parent.isEmpty()) {
+        batchSize = batchSize / 2 + 1;
```

Comment:
I noticed that very occasionally when I request too many bodies at the same time, some peers do not give me back all the bodies back. I ended up in an endless loop. 
1. Give me 200 bodies
2. Peer gives back bodies without one
3. I import all blocks until the one that was not given back.
4. Repeat 1,2 forever because the one not given back is always the same first block now.

I noticed that reducing the batch size (to eventually ask just for 1 body) resolved that issue for me. I am happy to apply a different tactic.